File: Haskell/Day07.hs
import AdventHelper

import Data.List.Split

crabEnergy :: [Int] -> Int -> Int
crabEnergy xs p = sum $ map (\x -> abs (x - p)) xs

crabEnergy2 :: [Int] -> Int -> Int
crabEnergy2 xs p = sum $ map (\x -> mv (abs (x - p))) xs
  where mv n = n * (n + 1) `div` 2 

main = do
  putStrLn "Day 7"
  f <- readFile "../input/input07.txt"
  let ss = lines f
  let cs = map read $ splitOn "," $ head ss :: [Int]

  printSoln 1 $ minimum $ map (crabEnergy cs) $ range ((minimum cs), (maximum cs))
  printSoln 2 $ minimum $ map (crabEnergy2 cs) $ range ((minimum cs), (maximum cs))

File: Go/Day07.go
package main

import AH "./adventhelper"

import (
	"strconv"
	"strings"
)

func crabEnergy(cs []int, p1 bool) (energy int) {
	energy = (1 << 31) // a big number 
	hi, lo := AH.MaxAndMin(cs)

	for i := lo; i <= hi; i++ {
		temp_energy := 0
		for _, v := range cs {
			n := AH.AbsInt(v - i)
			if p1 {
				temp_energy += n
			} else {
				temp_energy += (n * (n + 1)) / 2
			}
		}

		if temp_energy < energy {
			energy = temp_energy
		}
	}

	return
}

func main() {
	ss, _ := AH.ReadStrFile("../input/input07.txt")
	ps := strings.Split(ss[0], ",")
	crabs := []int{}
	for _, s := range ps {
		c, _ := strconv.Atoi(s)
		crabs = append(crabs, c)
	}

	AH.PrintSoln(7, crabEnergy(crabs, true), crabEnergy(crabs, false))

	return
}
